library/java/net/openid/appauth/AuthorizationService.java

```diff
@@ -559,4 +577,120 @@ protected void onPostExecute(JSONObject json) {
+            InputStream is = null;
+            try {
+                Uri uri = (mResponse.request.configuration.discoveryDoc.getJwksUri());
+                HttpURLConnection conn =
+                        mClientConfiguration.getConnectionBuilder()
+                        .openConnection(uri);
+                conn.setRequestMethod("GET");
+                //conn.setDoOutput(true);
+
+                is = new BufferedInputStream(conn.getInputStream());
+                // readStream(in);
+                String response = Utils.readInputStream(is);
+                return new JSONObject(response);
+            } catch (IOException ex) {
+                Logger.debugWithStack(ex, "Failed to complete token validation request");
+                mException = AuthorizationException.fromTemplate(
+                    GeneralErrors.NETWORK_ERROR, ex);
+            } catch (JSONException ex) {
+                Logger.debugWithStack(ex, "Failed to complete token validation request");
+                mException = AuthorizationException.fromTemplate(
+                    GeneralErrors.JSON_DESERIALIZATION_ERROR, ex);
+            } finally {
+                Utils.closeQuietly(is);
+            }
+            return null;
+        }
+
+        @Override
+        protected void onPostExecute(JSONObject json) {
+            JSONObject requiredJson = null;
+
+            try {
+                String[] split = mResponse.idToken.split("\\.");
+
+                String decodeTokenHeader = Utils.decodeBase64url(split[0]);
+                String decodeTokenBody = Utils.decodeBase64url(split[1]);
+
+                JSONObject jsonObjectHeader = new JSONObject(decodeTokenHeader);
+                JSONObject jsonObjectBody = new JSONObject(decodeTokenBody);
+
+                JSONArray jsonArray = json.getJSONArray("keys");
+
+                for (int i = 0; i < jsonArray.length(); i++) {
+                    if (jsonArray.getJSONObject(i).optString("alg")
+                            .contains(jsonObjectHeader.getString("alg"))) {
+                        requiredJson = jsonArray.getJSONObject(i);
+                        break;
+                    }
+                }
+
+                Logger.debug("Token validation with %s completed",
+                        this.mResponse.request.configuration
+                            .discoveryDoc.getJwksUri());
+
+                if (mException != null
+                        || requiredJson == null
+                        || !json.optString("error", "").equals("")
+                        || !jsonObjectHeader.getString("kid").equals(requiredJson.getString("kid"))
+                        || !jsonObjectBody.getString("aud").equals(mResponse.request.clientId)) {
+                    mCallback.onTokenValidationRequestCompleted(false, mException);
+                    return;
+                }
+
+                Calendar calendar = Calendar.getInstance();
```

Comment:
I'm also not convinced that you're gaining much from calendar beyond directly using the exp value. Code can probably just be:

```java
long expirationTime = Long.parseLong(jsonObjectBody.getString("exp") * 1000L;
if (expirationTime > mClock.getCurrentTimeMillis()) { ... } else { ... }
```